unbit/uwsgi
# plugins/php/session.c: for php7 return failures only on failure

Hello,

One of the changes in php7 listed in http://php.net/manual/en/migration70.incompatible.php was change to the custom session handler. Since php7 return values from session handlers are subjected to stricter checking.

According to https://bugs.php.net/bug.php?id=71187 

> [2016-01-15 21:26 UTC] yohgaki@php.net
> Please note that
> 
>  - PHP 7.0 and up does not allow buggy return values from user save handler.
> user read handler MUST return "string" data for success always.
>  - Native save handler must return SUCCESS for successful cases including non-existing session data. 
>  - FALSE/failure means "Something wrong in read" such as permission/network/etc errors.

How to test this patch:

1. build uwsgi with `python uwsgiconfig.py --clean --build php.ini`
2. prepare test script:
```
$ mkdir /tmp/uwsgi-testing
$ cat /tmp/uwsgi-testing/index.php 
<?php

session_start();
$_SESSION['t'] = 't';
error_log(var_export(session_id(), true));
session_commit();

session_start();
error_log(var_export(session_id(), true));
session_regenerate_id();
error_log(var_export(session_id(), true));
session_commit();

session_start();
session_destroy();
echo 'done';
```
3. execute uwsgi with `./uwsgi --http-socket 127.0.0.1:8080 --route '.* uwsgi:,14,0' --cache2 'name=session,items=1000,store=/tmp/uwsgi-session-cache,bitmap=1' --php-set 'session.save_handler=uwsgi' --php-set 'session.save_path=session' --php-docroot /tmp/uwsgi-testing/`
4. test it with `curl -v http://127.0.0.1:8080/index.php`

What should you see in the uwsgi log with this patch:

```
spawned uWSGI worker 1 (and the only) (pid: 585303, cores: 1)
'jqtntqgksj9ctk2f5oorpu2jpi'
'jqtntqgksj9ctk2f5oorpu2jpi'
'njaq8uq9dspjkhs0000kvn5q2d'
[pid: 585303|app: -1|req: -1/1] 127.0.0.1 () {24 vars in 267 bytes} [Wed Sep 20 14:59:22 2017] GET /index.php => generated 4 bytes in 0 msecs (HTTP/1.1 200) 6 headers in 252 bytes (0 switches on core 0)
'ecge1hkbs5f2u8i2n9cr79jlm2'
'ecge1hkbs5f2u8i2n9cr79jlm2'
'r052h31qm5u4gsj4ko4i5o5vk1'
[pid: 585303|app: -1|req: -1/2] 127.0.0.1 () {24 vars in 267 bytes} [Wed Sep 20 14:59:24 2017] GET /index.php => generated 4 bytes in 0 msecs (HTTP/1.1 200) 6 headers in 252 bytes (0 switches on core 0)
```

And without it:

```
spawned uWSGI worker 1 (and the only) (pid: 584500, cores: 1)
PHP Warning:  session_start(): Failed to read session data: uwsgi (path: session) in /tmp/tmp.SrWfLnAzQb/index.php on line 3
'ci7cfmg837jq9tqcma3qds2lub'
PHP Warning:  session_start(): Failed to read session data: uwsgi (path: session) in /tmp/tmp.SrWfLnAzQb/index.php on line 8
'ci7cfmg837jq9tqcma3qds2lub'
PHP Warning:  session_regenerate_id(): Cannot regenerate session id - session is not active in /tmp/tmp.SrWfLnAzQb/index.php on line 10
'ci7cfmg837jq9tqcma3qds2lub'
PHP Warning:  session_start(): Failed to read session data: uwsgi (path: session) in /tmp/tmp.SrWfLnAzQb/index.php on line 14
PHP Warning:  session_destroy(): Trying to destroy uninitialized session in /tmp/tmp.SrWfLnAzQb/index.php on line 15
[pid: 584500|app: -1|req: -1/1] 127.0.0.1 () {24 vars in 267 bytes} [Wed Sep 20 14:58:31 2017] GET /index.php => generated 4 bytes in 0 msecs (HTTP/1.1 200) 3 headers in 142 bytes (0 switches on core 0)
PHP Warning:  session_start(): Failed to read session data: uwsgi (path: session) in /tmp/tmp.SrWfLnAzQb/index.php on line 3
'u6sjg9tdsd068cj1phkje21t2v'
PHP Warning:  session_start(): Failed to read session data: uwsgi (path: session) in /tmp/tmp.SrWfLnAzQb/index.php on line 8
'u6sjg9tdsd068cj1phkje21t2v'
PHP Warning:  session_regenerate_id(): Cannot regenerate session id - session is not active in /tmp/tmp.SrWfLnAzQb/index.php on line 10
'u6sjg9tdsd068cj1phkje21t2v'
PHP Warning:  session_start(): Failed to read session data: uwsgi (path: session) in /tmp/tmp.SrWfLnAzQb/index.php on line 14
PHP Warning:  session_destroy(): Trying to destroy uninitialized session in /tmp/tmp.SrWfLnAzQb/index.php on line 15
[pid: 584500|app: -1|req: -1/2] 127.0.0.1 () {24 vars in 267 bytes} [Wed Sep 20 14:58:34 2017] GET /index.php => generated 4 bytes in 0 msecs (HTTP/1.1 200) 3 headers in 142 bytes (0 switches on core 0)
```

Frequently php applications are calling `set_error_handler()` with custom function that will raise an exception for these warnings, making uwsgi session handler useless with php7.
